일기 CRUD API

pear96 opened the pull request:
일기 CRUD API 구현
- 다이어리 전체 조회시 페이지네이션 적용
- 입력값 유효성 검사
- 커스텀 예외 처리

close #2 #8 #11 #18

pear96 committed:
feat: 다이어리 전체 조회 페이지네이션 적용 #8

- userId 기준으로, 자신이 작성한 일기를 최신순으로 정렬해서 페이지로 받아온다.

pear96 committed:
feat: 유효성 검사 처리 #11

- ValidStringLength : 공백 포함한 실제 문자 길이 측정
- ValidTitle, ValidContent는 우선 길이 제한만 적용. 길이 제한을 위해 ValidStringLength를 가져와서 검사한다.
- ValidMongoId : MongoId의 형태에 맞게 검사한다.

pear96 committed:
feat: 커스텀 예외 처리 #18

- common에 예외 응답(`ErrorResponse`)과 예외(`ServiceException`)을 추가. `BaseErrorCode`는 enum 클래스의 형태 유지를 위해 만들었다.
- Emotion에 enum값에 속하지 않는 값이 들어올 경우 ServiceException을 던진다. 대신 이걸 받기 위해선 `HttpMessageNotReadableException → ValueInstantiationException → ServiceException` 을 타고 들어가서 처리해야한다
- 서버마다 예외처리 방식이 다르므로, api(mvc) 서버에 예외 코드와 예외 핸들러를 구현했다.

pear96 committed:
fix: StringLengthValidator 조건 수정

- null일 경우 valid 하지 않음으로 수정

pear96 committed:
fix: MongoIdValidator 조건 수정

- null일 경우 valid 하지 않음으로 수정

kochungcheon commented on diary-api/src/main/java/chungha/diary/util/validation/validator/StringLengthValidator.java:
```diff
@@ -0,0 +1,26 @@
+			return false;
+		}
+		// 코드 포인트 수를 기준으로 글자 수를 측정 (공백 포함)
+		int actualLength = value.codePointCount(0, value.length());
```
codePointCount 쓰신 이유가 궁금합니다

pear96 replied:
백엔드에서 글자 수 계산 시 특수 문자 고려, 프론트에서 입력 시 실시간 글자 수 표시 db 컬럼에는 여유롭게 600 ~ 700 (기술적 고려사항에 적혀있는 부분)

- 코드 유닛 : 컴퓨터가 저장할 수 있도록 쪼개 놓은 단위.
  - 기본 다국어 영역 (U+0000 ~ U+FFFF) , 추가 문자 영역 (U+10000 이상) => 2개 코드 유닛 사용
- 유니코드 포인트 : 하나의 문자에 대한 하나의 코드 포인트
  ex) A = U+0041, 가 = U+AC00, 😊 = U+1F60A (U+D83D와 U+DE0A를 합친것(서러게이트 페어))

`codePointCount()` => 유니코드 코드 포인트 개수를 세는 함수. 사람이 인식하는 문자 하나하나 세는 방식
String.length()는 UTF-16 코드 유닛의 개수를 반환
대부분의 문자들은 하나의 코드 유닛(2byte)로 표현되지만, 일부 특수문자는 2개의 코드 유닛으로 표현됨. 따라서 😊 는 길이가 1이 아니라 2라고 인식되는 것임


결론은 😊을 글자수 1개로 쳐주기 위해서 사용하는 함수.
그렇게 500자로 제한해두고, 이모지가 최대 4 bytes라는 점에서 DB는 VARCHAR(2000)로 하라고 하신다~

kochungcheon commented on diary-api/src/main/java/chungha/diary/repository/DiaryRepository.java:
```diff
@@ -26,10 +27,17 @@ public void saveDiary(Diary diary) {
 
-	public Page<Diary> getAllDiary(Pageable pageable) {
-		Query query = new Query().with(pageable);
+	public PagedModel<Diary> getAllDiary(String userId, Pageable pageable) {
```
Entity를 그대로 반환 하지 말고, 다른 객체로 보내는 건 어떤 가요!

pear96 replied:
아주 지당하신 말씀입니다 제가 미처 놓친부분입니다 감사합니다 시정하겠습니다

pear96 committed:
refactor: 페이지네이션 시 Entity가 아닌 DTO로 반환

- `PagedResourcesAssembler` : links를 추가해서 클라이언트에서 page 요청을 좀 더 하기 쉽게 해줌. Hateos 라이브러리 추가
- `page`, `size`, `sort` 키워드를 자동으로 인식해서 Pageable 사용 가능

pear96 committed:
Merge branch 'develop' into feat/crud-diary
